Fix remaining mypy issues in "core" Airflow

ashb opened the pull request:
This fixes the remaining Mypy errors for "core" Airflow, i.e. excluding anything in a provider, or its tests

Related #19891

Before: Found 144 errors in 48 files (checked 2707 source files); Found 11 errors in 5 files

After: Found 102 errors in 29 files (checked 2707 source files); Found 2 errors in 1 file (checked 384 source files)


<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

potiuk committed:
IOBase/IO[str] in python doesn't adhere to Liskov Substitution Principle

Co-authored-by: Josh Fell <josh.d.fell@astronomer.io>

ashb committed:
Fix remaining mypy issues in "core" Airflow

ashb commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1140,9 +1140,9 @@ def _send_sla_callbacks_to_processor(self, dag: DAG):
-            Stats.gauge(f'pool.open_slots.{pool_name}', slot_stats[PoolSlotState.OPEN.value])
-            Stats.gauge(f'pool.queued_slots.{pool_name}', slot_stats[PoolSlotState.QUEUED.value])
-            Stats.gauge(f'pool.running_slots.{pool_name}', slot_stats[PoolSlotState.RUNNING.value])
+            Stats.gauge(f'pool.open_slots.{pool_name}', slot_stats["open"])
```
Mypy complained about this as `slot_stats` is a TypedDict, so I removed this and deleted the PoolSlotState, as this is the _only_ place this class is used.

uranusjr replied:
But it’s too stupid to know enum values are constants anyway LOL

potiuk replied:
https://github.com/python/mypy/issues/8722

kaxil commented on airflow/api/common/experimental/mark_tasks.py:
```diff
@@ -194,13 +196,15 @@ def get_subdag_runs(
 
             current_task = current_dag.get_task(task_id)
             if isinstance(current_task, SubDagOperator) or current_task.task_type == "SubDagOperator":
+                if TYPE_CHECKING:
+                    assert current_task.subdag
```
Why this?

ashb replied:
Because mypy doesn't/can't know that `isinstance(current_task, SubDagOperator)` means `current_task.subdag` will be non-None as a result, so that's what this assert gives us.

kaxil replied:
We could also probably do this _I think_ (not tested) -- as we have marked `SubDagOperator.subdag` of type `DAG`.

```python
current_task = cast(SubDagOperator, current_task)
```
or
```python
current_task.subdag = cast(DAG, current_task.subdag)
```

uranusjr replied:
The most generally accepted way to work around this is to actually use `assert` _without_ the if-guard; this is strictly for the linters, so `assert` being optimised in `-O` is totally a good thing. This is why Airflow’s “never use assert ever” rule has not sit well with me; assert exists for a reason, and _this is exactly it_.

potiuk replied:
Well. What could I say :) . I was outvoted with that one :). So now there are the two of us.

potiuk replied:
We can always ask for reassumption of the vote BTW if we have strong case.

ashb replied:
I'm more than happy to have the assert without the `if` -- I think the main point of avoiding asserts at runtime is that the code must continue to function if the assert was never run (and so for most people the easiest thing is to avoid them entirely)

kaxil commented on airflow/executors/base_executor.py:
```diff
@@ -55,7 +55,7 @@ class BaseExecutor(LoggingMixin):
     """
 
-    job_id: Optional[str] = None
+    job_id: Optional[int] = None
```
We pass this as a `str` (from Backfill and Manual triggers) and `int` from Scheduler

https://github.com/apache/airflow/blob/5980d2b05eee484256c634d5efae9410265c65e9/airflow/jobs/backfill_job.py#L807

https://github.com/apache/airflow/blob/5980d2b05eee484256c634d5efae9410265c65e9/airflow/cli/commands/task_command.py#L172

https://github.com/apache/airflow/blob/5980d2b05eee484256c634d5efae9410265c65e9/airflow/jobs/scheduler_job.py#L662

https://github.com/apache/airflow/blob/5980d2b05eee484256c634d5efae9410265c65e9/airflow/jobs/base_job.py#L56

ashb replied:
I did wonder if this might cause a problem. I guess it should be `int | str | None` then.

kaxil replied:
Yeah

uranusjr replied:
Changing this to `int | str | None` affects quite a few things to make the changes non-trivial, so I created a separate PR for easier review: astronomer/airflow#1481.

kaxil commented on airflow/models/taskinstance.py:
```diff
@@ -1286,7 +1287,7 @@ def check_and_change_state_before_execution(
 
     def _date_or_empty(self, attr: str):
-        result = getattr(self, attr, None)  # type: datetime
+        result = getattr(self, attr, None)  # type: Optional[datetime]
```
```suggestion
        result: Optional[datetime] = getattr(self, attr, None)
```

nit: Feels more cleaner

ashb replied:
I'll see if that works -- I just added the `Optional[]` without really thinking/examining too much.

ashb replied:
Yes, that works.

uranusjr committed:
Explicitly convert TI State to DR State

Instead of using force-cast, this leverages Enum's built-in feature,
providing additional validation on the value passed in.

uranusjr commented on airflow/decorators/task_group.py:
```diff
@@ -61,7 +61,7 @@ def __attrs_post_init__(self):
         return TaskGroup(**kwargs)
 
-    def __call__(self, *args, **kwargs) -> R:
+    def __call__(self, *args, **kwargs) -> Union[R, TaskGroup]:
```
I’m wondering maybe `TaskGroupDecorator` does not need to be a generic and `R` can all be replaced with `DependencyMixin`.

ashb replied:
```
@task_group
def foo() -> BashOperator
    return BashOperator(task_id='a')

bash_op = foo()
```

ashb replied:
It maybe doesn't _need_ to be, but it feels like it should be.

The "Union" is because I couldn't find a way  of doing an overload on a class-level generic.

THe ideal is: If R is None return, then the return type of call is TaskGroup, else it's R

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -2095,8 +2096,7 @@ def get_email_subject_content(self, exception: BaseException) -> Tuple[str, str,
         else:
-            jinja_context = self.get_template_context()
-            jinja_context.update(additional_context)
+            jinja_context = {**cast(dict, self.get_template_context()), **additional_context}
```
This is where the deprecations are from, I’ll fix it.

ashb replied:
Yeah, I'm on it already @uranusjr

ashb replied:
(I've got another case of this inside utils.helpers too0

uranusjr committed:
Pull in typing changes from GH-20742

These warn for me locally (not sure why they're not triggered elsewhere,
Mypy is mysterious). But these changes all make sense to me.

Also some trivial string concatenation improvements.

ashb committed:
fixup! Fix remaining mypy issues in "core" Airflow

uranusjr commented on airflow/providers_manager.py:
```diff
@@ -571,10 +571,10 @@ def _add_taskflow_decorator(self, name, decorator_class_name: str, provider_pack
             try:
                 existing = self._taskflow_decorators[name]
-                other_name = f'{existing.__module__}.{existing.__name}'
+                other_name = f'{existing.__module__}.{existing.__name__}'
```
wat

potiuk replied:
ups.

ashb committed:
Use context_merge function

And also correct type and implementation of it to support everything
dict.update does.

uranusjr commented on airflow/utils/operator_helpers.py:
```diff
@@ -164,8 +164,10 @@ def determine(
+        # We have a typestub that says Context is a TypedDict, but at runtime it's actually a custom
+        # MutableMapping class.
+        if self._wildcard and isinstance(self._kwargs, Context):  # type: ignore
+            return lazy_mapping_from_context(self._kwargs)  # type: ignore
```
What’s the error here? I was able to remove both `type: ignore` comments locally and have pre-commit pass.

ashb replied:
I got an error saying something to the effect of "isistance doesn't work on TypedDict".

uranusjr commented on airflow/serialization/serialized_objects.py:
```diff
@@ -975,7 +975,7 @@ def serialize_task_group(cls, task_group: TaskGroup) -> Optional[Dict[str, Any]]
-                else (DAT.TASK_GROUP, SerializedTaskGroup.serialize_task_group(child))
+                else (DAT.TASK_GROUP, SerializedTaskGroup.serialize_task_group(cast("TaskGroup", child)))
                 for label, child in task_group.children.items()
             },
```
This feels like a good candidate for poylmorphism. If we have add an abstract method `serialize_for_task_group` on `DAGNode`, we can do

```python
"children": {
    label: child.serialize_for_task_group()
    for label, child in task_group.children.items()
}
```

to both avoid the `cast`, and better prevent bugs if we add more `DAGNode` subclasses in the future.

uranusjr committed:
Abstract task group children serialization

This introduces a new function 'serialize_for_task_group' to DAGNode,
which should implement serialization logic when a DAGNode is put into a
task group. Since all nodes belonging to a DAG are either the root
group, or always in a parent task group, this guarantees a DAG's all
nodes can be correctly serialized.

uranusjr committed:
Comment and docstring improvements

ashb commented on airflow/utils/operator_helpers.py:
```diff
@@ -164,7 +164,7 @@ def determine(
         """Dump the kwargs mapping to unpack with ``**`` in a function call."""
-        if self._wildcard and isinstance(self._kwargs, Context):
+        if self._wildcard and is_context_instance(self._kwargs):
             return lazy_mapping_from_context(self._kwargs)
```
@uranusjr 

```
airflow/utils/operator_helpers.py:168: error: Argument 1 to
"lazy_mapping_from_context" has incompatible type "Mapping[str, Any]"; expected
"Context"
                return lazy_mapping_from_context(self._kwargs)
```

That is why I had this as a type: ignore

ashb replied:
So either we need a type ignore on this line, or a cast.

uranusjr replied:
Ah right because if you wrap `isinstance()` inside a function Mypy can no longer recognise it 🤦

uranusjr committed:
Use different name to avoid wrong type inference

ashb committed:
Fix tests

ashb committed:
fixup! Fix tests

uranusjr committed:
Stringify scheduler job ID in Kubernetes executor

The executor declaring int as job ID type is inconsistent to code using
it, which can pass in str. So the declaration is changed to use str
instead, and implementation is reviewed to make sure it works correctly.

ashb committed:
fixup! Stringify scheduler job ID in Kubernetes executor

potiuk commented on airflow/executors/base_executor.py:
```diff
@@ -55,7 +55,7 @@ class BaseExecutor(LoggingMixin):
     """
 
-    job_id: Optional[str] = None
+    job_id: Union[None, int, str] = None
```
Maybe:

```suggestion
    job_id: Optional[Union[int, str]] = None
```

A bit longer, but I think conceptually more consistent with other optionals.

potiuk commented on airflow/models/baseoperator.py:
```diff
@@ -1600,6 +1601,10 @@ def get_serialized_fields(cls):
 
         return cls.__serialized_fields
 
+    def serialize_for_task_group(self) -> Tuple[DagAttributeTypes, Any]:
```
nice

potiuk commented on airflow/utils/log/logging_mixin.py:
```diff
@@ -73,11 +73,23 @@ def supports_external_link(self) -> bool:
 
     encoding: None = None
 
+    def fileno(self) -> int:
```
Hmmm. I had the impression this has been here already :)

ashb replied:
I thought so too but it complained without it -- this was mostly around the use in StdoutRedirector which wants this to be `IO[str]`

potiuk commented on airflow/version.py:
```diff
@@ -22,7 +22,7 @@
     import importlib_metadata as metadata
 except ImportError:
-    from importlib import metadata
+    from importlib import metadata  # type: ignore
```
```suggestion
    from importlib import metadata  # type: ignore[attr-defined]
```

ashb replied:
You mean `no-redef` :)

potiuk commented on airflow/plugins_manager.py:
```diff
@@ -24,12 +24,12 @@
     import importlib_metadata
 except ImportError:
-    from importlib import metadata as importlib_metadata
+    from importlib import metadata as importlib_metadata  # type: ignore
```
```suggestion
    from importlib import metadata as importlib_metadata  # type: ignore[attr-defined]
```

potiuk commented on airflow/providers_manager.py:
```diff
@@ -571,10 +571,10 @@ def _add_taskflow_decorator(self, name, decorator_class_name: str, provider_pack
             except Exception:
                 # If problem importing, then get the value from the functools.partial
-                other_name = self._taskflow_decorators._raw_dict[name].args[0]
+                other_name = self._taskflow_decorators._raw_dict[name].args[0]  # type: ignore
```
Not sure which error is silenced here but we started to silence specific error codes recently: https://mypy.readthedocs.io/en/stable/error_code_list.html#error-code-list

ashb replied:
I'll add `show_error_codes=True` in to the config in this PR too

potiuk replied:
Ahhhhh YEAH!

potiuk commented on airflow/utils/operator_helpers.py:
```diff
@@ -164,7 +164,10 @@ def determine(
+        # Context is a TypedDict at lint time, and Mypy would complain it cannot
+        # be used in isinstance. But the call works at runtime since the type is
+        # actually implemented as a custom mapping, so we ignore the Mypy error.
+        if self._wildcard and isinstance(self._kwargs, Context):  # type: ignore
```
Might be good also to review other type-ignores to see if we can silence a specific error (watch-out for `misc`)

ashb replied:
Have gone through all the ones added in this PR and converted them to specific codes where possible.

potiuk reviewed (approved):
Few nits

ashb committed:
Ignore specific type of mypy typing errors where possible

ashb commented on airflow/utils/operator_helpers.py:
```diff
@@ -164,10 +164,7 @@ def determine(
-        # be used in isinstance. But the call works at runtime since the type is
-        # actually implemented as a custom mapping, so we ignore the Mypy error.
-        if self._wildcard and isinstance(self._kwargs, Context):  # type: ignore
+        if self._wildcard and isinstance(self._kwargs, Context):
```
Despite mypy complaining about this before _in this PR_ it's now not, at least not for me when I run `pre-commit run -a mypy` locally.

I want to see what number CI gets -- it should be 102 errors left for the first mypy job.

potiuk replied:
```
pre-commit run mypy --all-files 2>&1 | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g" | grep "error:" | wc

    105     292    8934
```

BTW. The numbers might not add up. It's quite possible that fixing a mypy error in "selective" check will generate more errors elsewhere (in providers for example)

ashb replied:
Confirmed it's fine now. I have _no_ idea what was going on before but this was def a problem.

potiuk replied:
Possibly other fix fixed it. Also while MyPy is stable and repeatable, it is somewhat "vulnerable" to small changes - it generates more or less loosely related errors after small changes that should not (at least obviously) trigger them - at least that's the impression I had sometimes.

ashb committed:
Fix rebase

uranusjr commented on airflow/models/taskmixin.py:
```diff
@@ -114,6 +115,14 @@ class DAGNode(DependencyMixin, metaclass=ABCMeta):
+        tg: Optional["TaskGroup"] = getattr(self, 'task_group', None)
+        if tg and tg.node_id and tg.prefix_group_id:
+            # "task_group_id.task_id" -> "task_id"
+            return self.node_id[len(tg.node_id) + 1 :]
```
```suggestion
            return self.node_id.split(".", 1)[-1]
```

ashb replied:
This doesn't feel safe -- there are cases where there can be a dot in the task_id. There _shouldn't_ be, but I cant guarantee it.

uranusjr replied:
Makes sense
